Add Merbridge blog - use ebpf to accelerate the Istio dataplane (#10834)
istio/istio.io: content/en/blog/2022/merbridge/index.md

---
title: "Merbridge - Accelerate your mesh with eBPF"
description: "Replacing iptables rules with eBPF allows transporting data directly from inbound sockets to outbound sockets, shortening the datapath between sidecars and services."
publishdate: 2022-03-07
attribution: "Kebe Liu (DaoCloud), Xiaopeng Han (DaoCloud), Hui Li (DaoCloud)"
keywords: [Istio,ebpf,iptables,sidecar]
---

> Currently, service mesh technologies, led by Istio, are attracting more and more attention from enterprises. The secret of Istio’s abilities in traffic management, security, observability and policy is all in the Envoy proxy. Istio uses Envoy as the "sidecar" to intercept service traffic, with the kernels `netfilter` packet filter functionality configured by iptables.

Reviewer: ```suggestion
The secret of Istio’s abilities in traffic management, security, observability and policy is all in the Envoy proxy. Istio uses Envoy as the "sidecar" to intercept service traffic, with the kernel's `netfilter` packet filter functionality configured by iptables.
```

> There are shortcomings in using iptables to perform this interception. Since `netfilter` is a highly versatile tool for filtering packets, several routing rules and data filtering processes are applied before reaching the destination socket. For example, from the network layer to the transport layer, `netfilter` will be used for processing for several times with the rules predefined, like `pre_routing`, `post_routing` and etc. When the packet becomes a TCP packet or UDP packet, and is forwarded to user space, some additional steps like packet validation, protocol policy processing and destination socket searching will be performed. When a sidecar is configured to intercept traffic, the original data path can become very long, since duplicated steps are performed several times.

Reviewer: ```suggestion
There are shortcomings in using iptables to perform this interception. Since netfilter is a highly versatile tool for filtering packets, several routing rules and data filtering processes are applied before reaching the destination socket. For example, from the network layer to the transport layer, netfilter will be used for processing for several times with the rules predefined, like `pre_routing`, `post_routing` and etc. When the packet becomes a TCP packet or UDP packet, and is forwarded to user space, some additional steps like packet validation, protocol policy processing and destination socket searching will be performed. When a sidecar is configured to intercept traffic, the original data path can become very long, since duplicated steps are performed several times.
```

Network connection is essentially socket communication. eBPF provides a function `bpf_msg_redirect_hash`, to directly forward the packets sent by the application in the inbound socket to the outbound socket. By entering the function mentioned before, developers can perform any logic to decide the packet destination. According to this characteristic, the datapath of packets can noticeably be optimized in the kernel.

The `sock_map` is the crucial piece in recording information for packet forwarding. When a packet arrives, an existing socket is selected from the `sock_map` to forward the packet to. As a result, we need to save all the socket information for packets to make the transportation process function properly. When there are new socket operations — like a new socket being created — the `sock_ops` function is executed.  The socket metadata is obtained and stored in the `sock_map` to be used when processing packets. The common key type in the `sock_map` is a "quadruple" of source and destination addresses and ports. With the key and the rules stored in the map, the destination socket will be found when a new packet arrives.

> ## The Merbridge approaches

Reviewer: ```suggestion
## The Merbridge approach
```

In the past two years, eBPF becomes a hot-trending technology, and many projects based on eBPF have been raised to the community. Projects like Cilium, [px.dev](http://px.dev) provides us with great cases of eBPF in observability and network packet processing. With eBPF’s `sockops` and `redir` capabilities, data packets can be processed efficiently by directly being transported from the inbound socket to the outbound socket. In Istio scenarios, it will be possible to use eBPF to replace iptables rules to accelerate the data-plane traffic by shorten the data-path.

Now that we have open sourced the Merbridge project, and by applying the following command to your Istio-managed cluster, you will directly get the eBPF ability to achieve network acceleration.

{{< text bash >}}
$ kubectl apply -f https://raw.githubusercontent.com/merbridge/merbridge/main/deploy/all-in-one.yaml
{{< /text >}}

{{< warning >}}

> Attention：Currently only support kernel version ≥ 5.15. Please check your kernel version to make sure eBPF functions properly.

Reviewer 2: ```suggestion
Attention：Currently only support kernel version ≥ 5.7. Please check your kernel version to make sure eBPF functions properly.
```

{{< /warning >}}

### Utilize eBPF `sockops` for performance optimization

Network connection is essentially socket communication. eBPF provides us with a function `bpf_msg_redirect_hash`, to directly forward the packets sent by the application in the inbound socket to the outbound socket, which can noticeably optimize the processing of the packets in the kernel.

Here the `sock_map` is the crucial part to record the socket rules because an existing socket connection needs to be selected from the `sock_map` according to the current data packet information. Therefore, the socket information needs to be stored to the map at the hook of `sockops` or somewhere else, and being provided according to the key formed by given information(usually quadruple).

## Approaches

Next, we will introduce the detailed design and implementation principles of Merbridge step by step according to the actual scene, to give you a preliminary understanding of Merbridge and eBPF.

### Istio sidecar traffic interception based on iptables

> {{< image link="./1.png" caption="Istio Sidecar Traffic Interception Based on Iptables" >}}

Reviewer 3: `1.png` should be translated to English?

Author: Thanks!

Reviewer 4: There is an English version for this image, see the English version [here](https://jimmysong.io/en/blog/sidecar-injection-iptables-and-traffic-routing/).

> As shown above, when external traffic accesses certain application’s ports, it will be intercepted by PREROUTING rule in iptables, forwarded to port 15001 of the sidecar container, and handed over to Envoy for processing(Like the red path of 1, 2, 3, 4 in the graph above).

Reviewer 2: ```suggestion
As shown above, when external traffic accesses certain application’s ports, it will be intercepted by PREROUTING rule in iptables, forwarded to port 15006 of the sidecar container, and handed over to Envoy for processing(Like the red path of 1, 2, 3, 4 in the graph above).
```

Envoy processes the traffic by using the policies the control plane issued. After finishing processing, the traffic will be sent to the actual container port of the application container.

> When the application tries to access other services, it will be intercepted by OUTPUT rule in iptables, and then be forwarded to port 15006(Which Envoy is listening to) of the sidecar container. (Like the red path of 9, 10, 11, 12, which is similar to the inbound traffic processing)

Reviewer 2: ```suggestion
When the application tries to access other services, it will be intercepted by OUTPUT rule in iptables, and then be forwarded to port 15001(Which Envoy is listening to) of the sidecar container. (Like the red path of 9, 10, 11, 12, which is similar to the inbound traffic processing)
```

### Inbound traffic processing

The processing of inbound traffic is basically similar to outbound traffic, with the only difference: revising the port of the destination to 15006.

However, it should be noted that since eBPF cannot take effect in the specified namespace like iptables, the change will be global, which means that if we use a Pod that is not originally managed by Istio, or an external IP address, serious problems will be encountered like the connection not being established at all.

As a result, we designed a tiny control plane here (deployed in DaemonSet), which watches all pods, similar to kubelet’s watching to the pod lists of the node, to write the pod IP addresses that have been injected into the sidecar to the `local_pod_ips` map.

When processing the inbound traffic, if the destination address is not in the map, we will not do anything to the traffic to make the traffic being processed more flexibly and simply.

Otherwise, the steps are the same as for outbound traffic.

{{< image link="./3.png" caption="Processing Inbound Traffic" >}}

> ### Same-node acceleration

Reviewer 5: Do we only support using this for pods on the same node?

{{< warning >}}
Envoy will retry when the request fails, and this retry process will only occur once, which means the subsequent requests will be super fast.
{{< /warning >}}

{{< image link="./4.png" caption="Same-node acceleration" >}}

### Connection relationship

Before applying eBPF using Merbridge, the data-path between pods is like:

{{< image link="./5.png" caption="Iptables's Data-path" >}}

After applying Merbridge(eBPF), the outbound traffic will skip many kernel modules like the graph below to improve the performance:

{{< image link="./6.png" caption="EBPF's Data-path" >}}

If two pods are on the same machine, the connection can even be faster:

{{< image link="./7.png" caption="EBPF's Datapath on The Same Machine" >}}

As above, using eBPF to process the traffic on the machine can greatly reduce the number of processes performed in the kernel, and the communication quality can be improved as well.

## Performance test

{{< warning >}}
Currently the tests below are basic, and not tested in production yet.
{{< /warning >}}

> The graph below shows the overall latency after using eBPF instead of iptables (lower is better):

Reviewer 6: Two questions here, what's your payload size in your testing and what performance tool did you use to do the test?

Reviewer 2: Each request about 200 bytes. and use `wrk` as performance tool.
